compiler/resolution/cullOverReferences.cpp

```diff
@@ -582,746 +582,848 @@ bool isRefOrTupleWithRef(Symbol* index, int tupleElement)
 }
 
 //
-// This function adjusts calls to functions that have both
-// a ref and non-ref version. These calls are represented with
-// a ContextCallExpr. Whether to use the ref or value/const-ref
-// version is determined based on how the result of the call is used.
-//
-// This pass studies all of these call-sites and determines if the
-// "by-reference" version is "necessary".  If it is not, then the
-// compiler attempts to switch in the "by-value" version. This requires
-// the insertion of a temporary value near the call site.
+// Context and helper functions for the pass `cullOverReferences`.
 //
-// The compiler must also determine whether the "by value" implementation
-// will have inserted an autoCopy.   If so then the compiler attaches
-// the AUTO_COPY/AUTO_DESTROY flags as necessary to enable the
-// callDestructors pass to operate correctly.
-//
-// This pass also concerns itself with lowering ArgSymbols with
-// INTENT_REF_MAYBE_CONST to INTENT_REF or INTENT_CONST_REF, depending
-// on whether the formal was "set" in the body of the function.
-//
-void cullOverReferences() {
-
-  /* A note about the structure of this pass:
 
-     While it's interprocedural, it shouldn't present a big problem for
-     separate compilation. The reason is that functions compiled into
-     a library should know whether or not they are setting their arguments.
-     Combining that fact with the plan to make the compiler produce a
-     "header file" in addition to "implementation files" as part of a library,
-     there is no problem.
-
-     (However, any interface describing a callable function / call-back
-      function will need to have some other default -- possibly blank
-      argument intent leads to compilation error for ref-if-modified types
-      like arrays).
+// TODO: Move this into `resolution/loopDetails.cpp` and adjust the function
+// `gatherLoopDetails` to use it.
+struct LoopInfo {
```

Comment:
It is sometimes necessary to use anonymous namespaces when declaring types meant to be used only within one .cpp file. It's not a thing you must do here but I wanted to make you aware of the potential for an issue. If another .cpp defines a (local) struct with the same name, there will be link errors, unless each uses `namespace { }` around the type definitions.

Reply:
Ok. In the future I would like to migrate this to be used in `gatherLoopDetails` at which point I could put it in a header file.

Reply:
Right, but the same thing applies to the new type just below.